src/main/java/com/gregtechceu/gtceu/integration/ae2/machine/MEStockingHatchPartMachine.java

```diff
@@ -118,33 +124,81 @@ public ManagedFieldHolder getFieldHolder() {
-            if (autoPull) {
-                refreshList();
-            }
             syncME();
         }
     }
 
     @Override
     protected void syncME() {
-        MEStorage networkInv = this.getMainNode().getGrid().getStorageService().getInventory();
-        for (ExportOnlyAEFluidSlot slot : aeFluidHandler.getInventory()) {
-            var config = slot.getConfig();
-            if (config != null) {
-                // Try to fill the slot
-                var key = config.what();
-                long extracted = networkInv.extract(key, Long.MAX_VALUE, Actionable.SIMULATE, actionSource);
-                if (extracted >= minStackSize) {
-                    slot.setStock(new GenericStack(key, extracted));
-                    continue;
+        // Refreshes what things are being watched
+        // TODO: Be smart about this.
+        // We only need to update what we listen to when said list changes, so in
+        // refreshList() or when it's changed in the UI
+        // For now, periodically testing is fine since this is cheap afs
+        if (watcher == null) return;
+        watcher.reset();
+        if (autoPull) {
+            watcher.setWatchAll(true);
+        } else {
+            watcher.setWatchAll(false);
+            for (ExportOnlyAEFluidSlot slot : this.aeFluidHandler.getInventory()) {
+                if (slot.getConfig() == null) continue;
+                watcher.add(slot.getConfig().what());
+            }
+        }
+    }
+
+    @Override
+    public void updateWatcher(IStackWatcher newWatcher) {
+        watcher = newWatcher;
+        this.syncME();
+    }
+
+    @Override
+    public void onStackChange(AEKey what, long amount) {
+        if (autoPull) {
+            if (!(what instanceof AEFluidKey itemKey)) return;
+
+            // TODO: this is expensive? maybe move this somewhere else?
+            boolean changed;
+            if (autoPullTest != null && !autoPullTest.test(new GenericStack(itemKey, amount))) {
+                topItems.remove(what);
+                changed = true;
+            } else {
+                changed = topItems.insert(what, amount);
+            }
+
+            if (!changed) return;
+            // TODO: only do this once all the stacks have changed? Is there any way to know?
+            syncListToHandler();
+        } else {
+            for (ExportOnlyAEFluidSlot slot : this.aeFluidHandler.getInventory()) {
+                if (slot.getConfig() == null) continue;
+                AEKey key = slot.getConfig().what();
+                if (key.equals(what)) {
+                    if (amount > 0) {
+                        slot.setStock(new GenericStack(key, amount));
+                    } else {
+                        slot.setStock(null);
+                    }
                 }
             }
-            slot.setStock(null);
         }
     }
 
+    private void syncListToHandler() {
+        int index = 0;
+        for (var entry : topItems) {
```

Comment:
See prior comment - also can this code be deduplicated?

Reply:
Not in a clean way I could find, because of the usage of the slots, the fact that both lists dont have a shared superclass, etc